collections/map.go

```diff
@@ -0,0 +1,97 @@
+package collections
+
+import (
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// NewMap returns a Map given a StoreKey, a Prefix and the relative value and key encoders.
+func NewMap[K, V any](
+	sk storetypes.StoreKey, prefix Prefix,
+	keyEncoder KeyEncoder[K], valueEncoder ValueEncoder[V],
+) Map[K, V] {
+	return Map[K, V]{
+		kc:     keyEncoder,
+		vc:     valueEncoder,
+		sk:     sk,
+		prefix: prefix.Bytes(),
+	}
+}
+
+// Map represents the basic collections object.
+// It is used to map arbitrary keys to arbitrary
+// objects.
+type Map[K, V any] struct {
+	kc KeyEncoder[K]
+	vc ValueEncoder[V]
+
+	sk     storetypes.StoreKey
+	prefix []byte
+}
+
+// Set maps the provided value to the provided key in the store.
+func (m Map[K, V]) Set(ctx StorageProvider, key K, value V) {
+	keyBytes := m.encodeKey(key)
+
+	valueBytes, err := m.vc.Encode(value)
+	if err != nil {
+		panic(err)
+	}
+
+	m.getStore(ctx).Set(keyBytes, valueBytes)
+}
```

Comment:
Should also return an error here rather than panicking.

Reply:
addressed